### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by ernestmc**
- Cleanup some code.
- Add unit tests for the `ClientStateMachine`.
- Add instructions for running unit tests using gradle.

@basicNew can you please review?

### CODE REVIEW

File: src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java [hunk 1]
```diff
@@ -87,7 +89,7 @@ public static String translateState(int state) {
     }
   }
 
-  int latestGoalStatus;
+  public int latestGoalStatus;
```
Comment (basicNew) [inline src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java]: It is generally not a good idea to make an instance variable public just for the sake of running a test, as that violates encapsulation (I know we did that to quickly make the tests pass, but we should look for a workaround). I would suggest either:
- Find another way to write the test.
- If the previous is not possible, define a getter.

File: src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java [hunk 2]
```diff
@@ -453,15 +439,16 @@ public synchronized void transition(int goalStatus)
...
-      case ClientStates.PENDING:
-      case ClientStates.ACTIVE:
-        state = ClientStates.WAITING_FOR_CANCEL_ACK;
-        ret = true;
-        break;
-    }
+    boolean ret;
+    ArrayList<Integer> cancellableStates = new ArrayList<>(Arrays.asList(ClientStates.WAITING_FOR_GOAL_ACK,
+        ClientStates.PENDING, ClientStates.ACTIVE));
+
```
Comment (basicNew) [inline src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java]: Note that this should be the same:

```
bolean shouldCancel = cancellableStates.contains(state);
if (shouldCancel) {
state = ClientStates.WAITING_FOR_CANCEL_ACK;
}
return shouldCancel;
```

File: src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java [hunk 3]
```diff
@@ -1,37 +1,162 @@
...
+      actualState = subject.getState();
+      
+      // Assertion - Postcondition
+      assertEquals(expectedState, actualState);
   }
-
+  
+  @Test
+  public void testSetState() {
+       // Setup
```
Comment (basicNew) [inline src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java]: I think we can remove these comments by now

File: src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java [hunk 4]
```diff
@@ -1,37 +1,162 @@
...
+      checkCancelOnInitialNonCancellableState(ClientStateMachine.ClientStates.WAITING_FOR_RESULT);
+      checkCancelOnInitialNonCancellableState(ClientStateMachine.ClientStates.WAITING_FOR_CANCEL_ACK);
+      checkCancelOnInitialNonCancellableState(ClientStateMachine.ClientStates.RECALLING);
+      checkCancelOnInitialNonCancellableState(ClientStateMachine.ClientStates.PREEMPTING);
+      checkCancelOnInitialNonCancellableState(ClientStateMachine.ClientStates.DONE);
+      checkCancelOnInitialNonCancellableState(ClientStateMachine.ClientStates.LOST);
+  }  
+  
+  private void checkCancelOnInitialCancellableState(int initialState) {
+    subject.setState(initialState);
```
Comment (basicNew) [inline src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java:72]: indentation

**Comment by basicNew** [comment]
@ernestmc left a few comments, otherwise LGTM for a first take

File: src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java (hunk 1 above)
Comment (ernestmc) [reply src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java]: Defined a getter for this variable.

File: src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java (hunk 2 above)
Comment (ernestmc) [reply src/rosjava_actionlib/rosjava_actionlib/src/main/java/com/github/ekumen/rosjava_actionlib/ClientStateMachine.java]: Changed to that format.

File: src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java (hunk 3 above)
Comment (ernestmc) [reply src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java]: Unnecessary comments removed.

File: src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java (hunk 4 above)
Comment (ernestmc) [reply src/rosjava_actionlib/rosjava_actionlib/src/test/java/TestClientStateMachine.java:72]: Fixed indentation.

**Pull request merged**